R/names.R

```diff
@@ -0,0 +1,429 @@
+#'  universal names: ".y." "._z" "..2fa" ".FALSE"
+#' ```
+#'
+#' @seealso [rlang::names2()] returns the names of an object, after
+#'   making them `minimal`.
+#'
+#' The [Names attribute](https://principles.tidyverse.org/names-attribute.html)
+#' section in the "tidyverse package development principles".
+#'
+#' @examples
+#' # By default, `vec_as_names()` returns minimal names:
+#' vec_as_names(c(NA, NA, "foo"))
+#'
+#' # You can make them unique:
+#' vec_as_names(c(NA, NA, "foo"), repair = "unique")
+#'
+#' # Universal repairing fixes any non-syntactic name:
+#' vec_as_names(c("_foo", "+"), repair = "universal")
+#' @export
+vec_as_names <- function(names,
+                         ...,
+                         repair = c("minimal", "unique", "universal"),
+                         quiet = FALSE) {
+  if (!missing(...)) {
+    ellipsis::check_dots_empty()
+  }
+  repair <- validate_repair(repair)
+
+  if (is_function(repair)) {
+    names <- as_minimal_names(names)
+    names <- validate_minimal(repair(names), n = length(names))
+    return(names)
+  }
+
+  switch(arg_match(repair),
+    minimal = as_minimal_names(names),
+    unique = as_unique_names(as_minimal_names(names), quiet = quiet),
+    universal = as_universal_names(as_minimal_names(names), quiet = quiet)
+  )
+}
+
+validate_repair <- function(repair) {
+  if (is_formula(repair, scoped = TRUE, lhs = FALSE)) {
+    repair <- as_function(repair)
+  }
+
+  if (is_function(repair)) {
+    repair
+  } else {
+    arg_match(repair, c("minimal", "unique", "universal"))
+  }
+}
+validate_minimal <- function(names, n = NULL) {
+  if (is.null(names)) {
+    abort("Names repair functions can't return `NULL`.")
+  }
+  if (!is_character(names)) {
+    abort("Names repair functions must return a character vector.")
+  }
+  if (!is_null(n) && length(names) != n) {
+    abort(sprintf(
+      "Repaired names have length %d instead of length %d.",
+      length(names), n
+    ))
+  }
+  if (anyNA(names)) {
+    abort("Names repair functions can't return `NA` values.")
+  }
+  names
+}
+
+vec_names <- function(x,
+                      ...,
+                      repair = c("minimal", "unique", "universal"),
+                      quiet = FALSE) {
+  if (!missing(...)) {
+    ellipsis::check_dots_empty()
+  }
+  repair <- validate_repair(repair)
+
+  if (is_function(repair)) {
+    names <- minimal_names(x)
+    names <- validate_minimal(repair(names), n = length(names))
+    return(names)
+  }
+
+  switch(arg_match(repair),
+    minimal = minimal_names(x),
+    unique = as_unique_names(minimal_names(x), quiet = quiet),
+    universal = as_universal_names(minimal_names(x), quiet = quiet)
+  )
+}
+vec_repair_names <- function(x,
+                             repair = c("minimal", "unique", "universal"),
+                             ...,
+                             quiet = FALSE) {
+  set_bare_names(x, vec_names(x, ..., repair = repair, quiet = quiet))
+}
+
+minimal_names <- function(x) {
```

Comment:
Should we make these `names_minimal` etc?

Reply:
Then it would be strange to have `as_minimal_names()`, and `as_names_minimal()` looks strange as well? These are only used internally and in tests though, so perhaps not worth worrying about.